src/app/devices/devices.service.spec.ts

```diff
@@ -4,24 +4,407 @@
 **********************************************************************/
 import { TestBed } from '@angular/core/testing'
 import { Router } from '@angular/router'
+import { of, throwError } from 'rxjs'
+import { environment } from 'src/environments/environment'
 import { AuthService } from '../auth.service'
 
 import { DevicesService } from './devices.service'
 
 describe('DevicesService', () => {
```

Comment:
getHardwareINformation tests are missing

url condition for actions >= 100 is missing

Reply:
add test cases for  getHardwareinformation and actions >= 100

Reply:
done